File: inverse_kinematics/scripts/inverse_kinematics/YoubotKDL.py
#! /usr/bin/env python

import rospy
from math import pi
from sensor_msgs.msg import JointState
from geometry_msgs.msg import TransformStamped
import PyKDL
import tf2_ros

from YoubotKinematics import YoubotKinematics
from tf_conversions import posemath

import tf2_kdl


class YoubotKDL(YoubotKinematics):
    def __init__(self):
        super(YoubotKDL, self).__init__()

        # PyKDL
        self.kine_chain = PyKDL.Chain()
        self.setup_kdl_chain()
        self.current_joint_position = PyKDL.JntArray(self.kine_chain.getNrOfJoints())
        self.current_pose = PyKDL.Frame()

        self.fk_solver = PyKDL.ChainFkSolverPos_recursive(self.kine_chain)
        self.ik_solver = PyKDL.ChainIkSolverPos_LMA(self.kine_chain)
        self.jac_calc = PyKDL.ChainJntToJacSolver(self.kine_chain)

    def setup_kdl_chain(self):
        self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Joint.RotZ), PyKDL.Frame().DH(0.033, pi / 2, 0.147, 0)))
        self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Joint.RotZ), PyKDL.Frame().DH(0.155, 0, 0, pi / 2)))
        self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Joint.RotZ), PyKDL.Frame().DH(0.135, 0, 0, 0)))
        self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Joint.RotZ), PyKDL.Frame().DH(0.0, pi / 2, 0, pi / 2)))
        self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Joint.RotZ), PyKDL.Frame().DH(0, 0, 0.218, pi)))

    def joint_state_callback(self, msg):
        # Copies joint position into KDL JntArray
        for i in range(0, 5):
            self.current_joint_position[i] = msg.position[i]

    def forward_kinematics(self, joint, pose):
        self.fk_solver.JntToCart(joint, pose)

    def broadcast_pose(self, pose):
        trans = TransformStamped()

        tf2_kdl.do_transform_frame(pose, trans)
        trans.header.frame_id = "arm_link_0"
        trans.header.stamp = rospy.Time.now()
        trans.child_frame_id = "arm_end_effector"

        self.pose_broadcaster.sendTransform(trans)

    def get_jacobian(self, joint):
        jac = PyKDL.Jacobian(self.kine_chain.getNrOfJoints())
        self.jac_calc.JntToJac(joint, jac)
        return jac

    def inverse_kinematics_closed(self, desired_pose):

        required_joint = PyKDL.JntArray(self.kine_chain.getNrOfJoints())
        # Current joint array, desired frame, required joint
        self.ik_solver.CartToJnt(self.current_joint_position, desired_pose, required_joint)

        return required_joint

    def run(self):
        rate = rospy.Rate(200)
        while not rospy.is_shutdown():
            # Find your current cartesian pose
            self.forward_kinematics(self.current_joint_position, self.current_pose)
            self.broadcast_pose(self.current_pose)

            # Ikine

            rate.sleep()



if __name__ == '__main__':
    rospy.init_node('youbot')
    youbot = YoubotKDL()

    youbot.run()

File: inverse_kinematics/scripts/YoubotKDL.cpp
#include <inverse_kinematics/YoubotKDL.h>

int YoubotKDL::init()
{

    setup_kdl_chain();
    current_joint_position = KDL::JntArray(kine_chain.getNrOfJoints());

    KDL::ChainFkSolverPos_recursive fk_solver = KDL::ChainFkSolverPos_recursive(kine_chain);
    KDL::ChainIkSolverPos_LMA ik_solver = KDL::ChainIkSolverPos_LMA(kine_chain);
    KDL::ChainJntToJacSolver jac_solver = KDL::ChainJntToJacSolver(kine_chain);

    ros::Subscriber joint_sub = n.subscribe<sensor_msgs::JointState>("/joint_states", 1, &YoubotKDL::joint_state_callback,
                                                                                                     this);
}

KDL::Jacobian YoubotKDL::get_jacobian(KDL::ChainJntToJacSolver jac_solver)
{

    KDL::Jacobian jac = KDL::Jacobian(kine_chain.getNrOfJoints());
    jac_solver.JntToJac(current_joint_position, jac);
    return jac;
}

int YoubotKDL::forward_kinematics(KDL::ChainFkSolverPos_recursive fk_solver, KDL::JntArray current_joint_position)
{
    return fk_solver.JntToCart(current_joint_position, current_pose, 6);
}

void YoubotKDL::joint_state_callback(const sensor_msgs::JointState::ConstPtr &q)
{
    current_joint_position.data(0) = 0.0;
    for (int i = 0; i < 5; i++)
        current_joint_position.data(i + 1) = q->position.at(i);
}

void YoubotKDL::setup_kdl_chain()
{
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH(-0.024,   M_PI,  0.096, 160*M_PI/180)));
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH( 0.033, M_PI_2, -0.019, M_PI + (169 * M_PI / 180.0))));
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH(-0.155,      0,      0, M_PI_2 + (-65.0 * M_PI / 180.0))));
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH(-0.135,      0,      0, (146 * M_PI / 180.0))));
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH( 0.002, M_PI_2,      0, M_PI_2 + (-102.5 * M_PI / 180.0))));
    kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH(     0,   M_PI,  -0.21, -150*M_PI/180+(167.5 * M_PI / 180.0))));
}

KDL::JntArray YoubotKDL::inverse_kinematics_closed(KDL::JntArray current_joint_position, KDL::Frame desired_pose, KDL::ChainIkSolverPos_LMA ik_solver)
{
    KDL::JntArray required_joint = KDL::JntArray(kine_chain.getNrOfJoints());
    ik_solver.CartToJnt(current_joint_position, desired_pose, required_joint);

    return required_joint;
}

void YoubotKDL::broadcast_pose()
{
    ros::Rate r(200);

    while (n.ok())
    {

        tr_stamped = tf2::kdlToTransform(current_pose);

        trans.setOrigin(tf::Vector3(tr_stamped.transform.translation.x,
                                    tr_stamped.transform.translation.y,
                                    tr_stamped.transform.translation.z));

        q.setW(tr_stamped.transform.rotation.w);
        q.setX(tr_stamped.transform.rotation.x);
        q.setY(tr_stamped.transform.rotation.y);
        q.setZ(tr_stamped.transform.rotation.z);

        trans.setRotation(q);

        br.sendTransform(tf::StampedTransform(trans, ros::Time::now(), "base_link", "arm_end_effector"));
        r.sleep();

        ros::spinOnce();
    }

    return 0;
}